wasm/src/programs/snapshot_query.rs

```diff
@@ -0,0 +1,202 @@
+            .map(|cm_s| fetch_state_path_at_root(node_url, cm_s.as_str(), root_str.as_str()));
+        let results = join_all(futs).await;
+
+        // 3) Insert all paths; verify consistency against the pinned root.
+        for (cm, res) in commitments.iter().zip(results.into_iter()) {
+            let sp_str = res?;
+            // Optional safety: parse and ensure the path's root matches the pinned root.
+            let sp = StatePath::<CurrentNetwork>::from_str(&sp_str).map_err(|e| anyhow!(e.to_string()))?;
+            let path_root = sp.global_state_root().to_string();
+            if path_root != snap_root {
+                // Strategy: bail and let caller retry; or resnapshot + refetch here.
+                bail!("State path root mismatch: expected {}, got {}", snap_root, path_root);
+            }
+            query.add_state_path(&cm.to_string(), &sp_str)?;
+        }
+
+        Ok(query)
+    }
+}
+
+#[async_trait::async_trait(?Send)]
+impl QueryTrait<CurrentNetwork> for SnapshotQuery {
+    fn current_state_root(&self) -> Result<<CurrentNetwork as Network>::StateRoot> {
+        Ok(self.state_root)
+    }
+
+    async fn current_state_root_async(&self) -> Result<<CurrentNetwork as Network>::StateRoot> {
+        Ok(self.state_root)
+    }
+
+    fn get_state_path_for_commitment(
+        &self,
+        commitment: &Field<CurrentNetwork>,
+    ) -> Result<StatePath<CurrentNetwork>> {
+        self.state_paths
+            .get(commitment)
+            .cloned()
+            .ok_or_else(|| anyhow!("State path not found for commitment"))
+    }
+
+    async fn get_state_path_for_commitment_async(
+        &self,
+        commitment: &Field<CurrentNetwork>,
+    ) -> Result<StatePath<CurrentNetwork>> {
+        self.state_paths
+            .get(commitment)
+            .cloned()
+            .ok_or_else(|| anyhow!("State path not found for commitment"))
+    }
+
+    fn current_block_height(&self) -> Result<u32> {
+        Ok(self.block_height)
+    }
+
+    async fn current_block_height_async(&self) -> Result<u32> {
+        Ok(self.block_height)
+    }
+}
+
+/* --------------------------- internal helpers --------------------------- */
+
+/// Heuristic: detect plaintext records from JS inputs and compute commitments.
+fn collect_commitments_from_inputs(
+    program_id: &ProgramIDNative,
+    record_name: &IdentifierNative,
+    view_key: Field<CurrentNetwork>,
+    js_inputs: &[JsValue],
+) -> anyhow::Result<Vec<Field<CurrentNetwork>>> {
+    let mut out = Vec::new();
+
+    for js in js_inputs {
+        if let Some(s) = js.as_string() {
+            // Quick filter: record plaintexts include a `_nonce` field.
+            if !s.contains("_nonce") {
+                continue;
+            }
+            // Try parse as a plaintext record. Skip if not a record.
+            if let Ok(rec) = RecordPlaintextNative::from_str(&s) {
+                let cm = rec.to_commitment(program_id, record_name, &view_key);
```

Comment:
```suggestion
                let record_view_key = (&**view_key * &self.nonce).to_x_coordinate();
                let commitment = record.to_commitment(program_id, record_name, &record_view_key);
```

Note here that the record view key is a different thing from the view key of the account. Each record has its own view key, the above is how you compute it.